### PR TITLE
Add "Move To" functionality to rearrange file tree

### PR DESCRIPTION
**Pull request by gideonthomas**
This uses the bramble-move-file extension found at https://github.com/gideonthomas/bramble-move-file to allow moving files in the file tree.

### CODE REVIEW

File: src/bramble/client/main.js [hunk 1]
```diff
@@ -513,6 +514,42 @@ define([
                 };
             }
 
+            function moveFile(filePath, to, callback) {
```
Comment (humphd) [inline src/bramble/client/main.js]: s/to/destDirPath/

File: src/bramble/client/main.js [hunk 2]
```diff
@@ -513,6 +514,42 @@ define([
...
+                        if(err) {
+                            return callback(err);
+                        }
+
+                        _fs.writeFile(newFilePath, data, function(err) {
+                            if(err) {
+                                return callback(err);
+                            }
+
+                            callback(null, newFilePath);
```
Comment (humphd) [inline src/bramble/client/main.js]: Why do you send back the path? You already know what it will be.

File: src/bramble/client/main.js [hunk 3]
```diff
@@ -571,8 +608,53 @@ define([
...
+                _fs.stat(src, function(err, stats) {
+                    if(err) {
+                        return callback(err);
+                    }
+
+                    if(stats.isFile()) {
+                        return moveFile(src, dest, callback);
+                    }
+
+                    shell.find(src, {
```
Comment (humphd) [inline src/bramble/client/main.js]: Rewrite this like so:

```js
function move(nodePath, next) {
    ...
}
shell.find(src, {exec: move}, callback);
```

File: src/bramble/client/main.js [hunk 4]
```diff
@@ -571,8 +608,53 @@ define([
...
+                                if(err) {
+                                    return next(err);
+                                }
+
+                                // Directory contains something
+                                if(contents.length > 0) {
+                                    return next();
+                                }
+
+                                moveEmptyDirectory(nodePath, destinationDirectory, next);
```
Comment (humphd) [inline src/bramble/client/main.js]: Nice work figuring out this edge case early.

File: src/filesystem/impls/filer/BracketsFiler.js [hunk 5]
```diff
@@ -140,6 +140,15 @@ define(function (require, exports, module) {
...
         watch: function(path, options, callback) {
             proxyCall("watch", {args: [path, options], persist: true}, callback);
+        },
+        ls: function(path, options, callback) {
+            proxyCall("ls", {args: [path, options], shell: true}, callback);
+        },
+        mv: function(src, dest, callback) {
+            proxyCall("mv", {args: [src, dest], shell: true}, callback);
+        },
+        rm: function(path, options, callback) {
```
Comment (humphd) [inline src/filesystem/impls/filer/BracketsFiler.js]: Thank you for adding this.

File: src/filesystem/impls/filer/BracketsFiler.js [hunk 6]
```diff
@@ -140,6 +140,15 @@ define(function (require, exports, module) {
         },
         watch: function(path, options, callback) {
             proxyCall("watch", {args: [path, options], persist: true}, callback);
+        },
```
Comment (humphd) [inline src/filesystem/impls/filer/BracketsFiler.js]: Should we remove the `mkdirp` above in this file and use the shell trick?

File: src/extensions/default/bramble-move-file/MoveToDialog.js [hunk 7]
```diff
@@ -0,0 +1,191 @@
...
+    var Dialogs                = brackets.getModule("widgets/Dialogs");
+    var DefaultDialogs         = brackets.getModule("widgets/DefaultDialogs");
+    var EditorManager          = brackets.getModule("editor/EditorManager");
+    var ProjectManager         = brackets.getModule("project/ProjectManager");
+    var CommandManager         = brackets.getModule("command/CommandManager");
+    var Commands               = brackets.getModule("command/Commands");
+    var LiveDevMultiBrowser    = brackets.getModule("LiveDevelopment/LiveDevMultiBrowser");
+    var KeyEvent               = brackets.getModule("utils/KeyEvent");
+    var StartupState           = brackets.getModule("bramble/StartupState");
+    var Filer                  = brackets.getModule("filesystem/impls/filer/BracketsFiler");
```
Comment (humphd) [inline src/extensions/default/bramble-move-file/MoveToDialog.js]: Call this `BracketsFiler` so it's clear which one we're using.

File: src/extensions/default/bramble-move-file/MoveUtils.js [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+            if(contents.indexOf(srcNodeName) !== -1) {
+                return deferred.reject({ type: NEEDS_RENAME });
+            }
+
+            fs.mv(src, dest, function(err) {
+                if(err) {
+                    return deferred.reject(err);
+                }
+
+                BlobUtils.preload(dest, function(err) {
```
Comment (humphd) [inline src/extensions/default/bramble-move-file/MoveUtils.js]: Do this after you're done changing the fs (i.e., after the `rm` below) as the last thing you do before you resolve.

File: src/utils/BrambleExtensionLoader.js [hunk 9]
```diff
@@ -33,7 +33,8 @@ define(function (require, exports, module) {
         "Autosave",
         "brackets-paste-and-indent",
         "BrambleUrlCodeHints",
-        "UploadFiles"
+        "UploadFiles",
+        "bramble-move-file"
```
Comment (humphd) [inline src/utils/BrambleExtensionLoader.js]: Update the main README file to include this in the list of extensions we load by default.

File: src/filesystem/impls/filer/BracketsFiler.js (hunk 6 above)
Comment (gideonthomas) [inline src/filesystem/impls/filer/BracketsFiler.js]: Yeah, I can do that in a separate patch

**Comment by gideonthomas** [comment]
I tried to expand the destination directory in the file tree when you move but there doesn't seem to be an easy programmatic way to do it coz of the "Working Set" rubbish. I'll try to fix in a follow up patch.

File: src/bramble/client/main.js [hunk 10]
```diff
@@ -571,8 +645,46 @@ define([
...
+            case "rm":
+                path = args[0];
+                shell.rm(path, args[1], function(err) {
+                    if(err) {
+                        return callback(err);
+                    }
+
+                    // If this is not a recursive `rm`, immediately trigger the
+                    // `fileDelete` event and call the callback
+                    if(!args[1] || !args[1].recursive) {
```
Comment (humphd) [inline src/bramble/client/main.js]: On line 666 do `var options = args[1]` and use that here.

**Comment by humphd** [comment]
Looks good.  One suggestion, r=me otherwise.

**Comment by sedge** [comment]
This landed! Woo!

**Pull request merged**